src/queries/role.ts

```diff
@@ -28,22 +29,73 @@ const matchAccessibleNameIfNeeded = (
   );
 };
 
+const accessibilityStates = [
+  'disabled',
+  'selected',
+  'checked',
+  'busy',
+  'expanded',
+] as const;
+
 const queryAllByRole = (
   instance: ReactTestInstance
 ): ((role: TextMatch, options?: ByRoleOptions) => Array<ReactTestInstance>) =>
   function queryAllByRoleFn(role, options) {
-    return instance.findAll(
-      (node) =>
+    return instance.findAll((node) => {
+      const matchRole =
         typeof node.type === 'string' &&
-        matchStringProp(node.props.accessibilityRole, role) &&
-        matchAccessibleNameIfNeeded(node, options?.name)
-    );
+        matchStringProp(node.props.accessibilityRole, role);
+
+      if (!matchRole) return false;
+
+      if (options?.name) {
+        if (!matchAccessibleNameIfNeeded(node, options.name)) {
+          return false;
+        }
+      }
+
+      return accessibilityStates.every((accessibilityState) => {
+        const queriedState = options?.[accessibilityState];
+
+        if (typeof queriedState !== 'undefined') {
```

Comment:
We might alter the order of the checks and early false exits from cheapest to most expensive to check. So checking role name is the cheapes, then checking accessiblity state, then accessibleName